### PR TITLE
Add Hudi connector

### PR DESCRIPTION
**Pull request by hoaihuongbk**
Apache Hudi is a rich platform to build streaming data lakes with incremental data pipelines on a self-managing database layer while being optimized for lake engines and regular batch processing. For more, see https://hudi.apache.org/

Changes:
- Add a Hudi connector to enable read/write from the Hudi table.

### CODE REVIEW

**Comment by JorisTruong** [comment]
Thanks for contributing @hoaihuongbk ! Will review as soon as possible and also #258

File: pom.xml [hunk 1]
```diff
@@ -98,6 +98,8 @@
         <maven.site.plugin.version>3.9.1</maven.site.plugin.version>
         <nexus.staging.maven.plugin.version>1.6.8</nexus.staging.maven.plugin.version>
         <typesafe.config.version>1.4.2</typesafe.config.version>
+        <hudi.version>0.11.0</hudi.version>
```
Comment (qxzzxq) [inline pom.xml:101]: It seems that there are some dependency resolution issues with Spark 2.3 and 2.4. 
You can add the compatible version to the corresponding profile (line 400, tag <profiles>)

File: pom.xml [hunk 2]
```diff
@@ -178,6 +180,18 @@
             <version>${delta.version}</version>
         </dependency>
 
+        <dependency>
+            <groupId>org.apache.hudi</groupId>
+            <artifactId>hudi-spark3.1-bundle_${scala.compat.version}</artifactId>
```
Comment (qxzzxq) [inline pom.xml:185]: use property to avoid hardcoded spark version

**Comment by qxzzxq** [comment]
Thanks for the PR. It looks good. There are some resolution issues during the test for Spark 2.3 and 2.4. 
In addition, can you add some test cases for the new connector?

**Comment by hoaihuongbk** [comment]
> Thanks for the PR. It looks good. There are some resolution issues during the test for Spark 2.3 and 2.4. In addition, can you add some test cases for the new connector?

Hudi only supports spark 2.4.x and later. So, how can I exclude it for the spark 2.3 profile? Any advice?

**Comment by qxzzxq** [comment]
It seems that Hudi 0.7 supports Spark 2.3 (https://hudi.apache.org/docs/0.7.0/quick-start-guide). Try this version or previous ones. If some v0.7+ features are used in the unit test, you can skip those tests.

**Comment by qxzzxq** [comment]
good job :D seems that all the tests have been passed. Can you add some simple test cases just to cover those lines mentioned in https://github.com/SETL-Framework/setl/pull/257/checks?check_run_id=6989087153 ?

Then I will merge the PR

**Comment by hoaihuongbk** [comment]
Added more test cases

**Review by qxzzxq — APPROVED** [review]

**Comment by qxzzxq** [comment]
Thanks for the PR! I will merge it into the main branch. The snapshot version (`1.0.0-SNAPSHOT`) should contain this PR. Then I will need to renew our GPG key (#251) before releasing the stable version. I will do it ASAP

**Comment by hoaihuongbk** [comment]
Great. Thanks for much @qxzzxq

**Pull request merged**